activerecord/test/cases/insert_all_test.rb

```diff
@@ -393,6 +393,66 @@ def test_upsert_all_uses_given_updated_on_over_implicit_updated_on
     assert_equal updated_on, Book.find(101).updated_on
   end
 
+  def test_upsert_all_sets_timestamps_on_creation_when_told_to_set_timestamps_on_create
+    # TODO: Should we skip this under any circumstances?
+
+    Book.upsert_all [{ id: 101, name: "Out of the Silent Planet", published_on: Date.new(1938, 4, 1) }], set_timestamps_on_create: true
+
+    book = Book.find(101)
+
+    assert_not_nil book.created_at
+    assert_not_nil book.updated_at
+    assert_not_nil book.updated_on
+
+    assert_in_delta Time.now, book.created_at, 1
+    assert_equal book.created_at, book.updated_at
+    assert_equal book.updated_at.to_date, book.updated_on
+  end
+
+  def test_upsert_all_does_not_touch_existing_timestamps_on_identical_record_when_told_to_set_timestamps_on_create
+    # TODO: Should we skip this under any circumstances?
+
+    created_at = Time.now.utc - 30
+    Book.create! id: 101, name: "Out of the Silent Planet", published_on: Date.new(1938, 4, 1), created_at: created_at, updated_at: created_at, updated_on: created_at.to_date
+
+    Book.upsert_all [{ id: 101, name: "Out of the Silent Planet", published_on: Date.new(1938, 4, 1) }], set_timestamps_on_create: true
+
+    book = Book.find(101)
+
+    assert_in_delta created_at, book.created_at, 1
+    assert_equal book.created_at, book.updated_at
+    assert_equal book.updated_at.to_date, book.updated_on
+  end
+
+  def test_upsert_all_does_not_touch_creation_timestamps_on_updated_record_when_told_to_set_timestamps_on_create
+    # TODO: Should we skip this under any circumstances?
+
+    created_at = Time.now.utc - 30
+    Book.create! id: 101, name: "Out of the Silent Planet", published_on: Date.new(1938, 4, 1), created_at: created_at, updated_at: created_at, updated_on: created_at.to_date
+
+    Book.upsert_all [{ id: 101, name: "Out of the Silent Planet", published_on: Date.new(1938, 4, 8) }], set_timestamps_on_create: true
+
+    book = Book.find(101)
+
+    assert_in_delta created_at, book.created_at, 1
+    assert_not_equal book.created_at, book.updated_at
+    assert_equal book.updated_at.to_date, book.updated_on
+  end
+
+  def test_upsert_uses_explicit_timestamps_even_if_told_to_set_timestamps_on_create
+    # TODO: Should we skip this under any circumstances?
+
+    created_at = Time.now.utc - 30
+
+    Book.upsert_all [{ id: 101, name: "Out of the Silent Planet", published_on: Date.new(1938, 4, 8), created_at: created_at, updated_at: created_at, updated_on: created_at.to_date }], set_timestamps_on_create: true
+
+    book = Book.find(101)
+
+    assert_in_delta created_at, book.created_at, 1
+    assert_equal book.created_at, book.updated_at
+    assert_equal book.updated_at.to_date, book.updated_on
+  end
```

Comment:
These tests all exercise the API through `upsert_all`. Is that sufficient, or should we test its usage in the various other methods (e.g. `insert_all`)?